sub-packages/bionemo-scdl/src/bionemo/scdl/io/single_cell_memmap_dataset.py

```diff
@@ -494,6 +522,110 @@ def load_h5ad(
 
         # Store the row idx array
         self.row_index[0 : num_rows + 1] = count_data.indptr.astype(int)
+        return adata.var, num_rows
+
+    def _lazy_load_h5ad(
+        self,
+        anndata_path: str,
+    ) -> Tuple[pd.DataFrame, int]:
+        """Method for lazy loading a larger h5ad file and converting it to the SCDL format.
+
+        This should be used in the case when the entire anndata file cannot be loaded into memory.
+        The anndata is loaded into memory lazy_load_block_size number of rows at a time. Each chunk
+        is converted into numpy memory maps which are then concatenated together.
+
+        Raises:
+            NotImplementedError if the data is not lazy loaded in the correct format.
+
+        Returns:
+            pd.DataFrame: var variables for features
+            int: number of rows in the dataframe.
+        """
+        column_mem_map_list = []
+        data_mem_map_list = []
+
+        adata = ad.read_h5ad(anndata_path, backed=True)
+        if not isinstance(adata.X, ad.experimental.CSRDataset):
+            raise NotImplementedError("Sparse Matrix cannot be lazy-loaded.")
+
+        num_rows = adata.X.shape[0]
+        mode = Mode.CREATE_APPEND.value
+
+        # Read the row indices into a memory map. This is done chunk by chunk.
+        self.row_index = _create_row_memmaps(num_rows, Path(self.data_path), mode, self.dtypes)
```

Comment:
This could be a bit annoying to refactor - we access self.dtypes when we are concatenating arrays for example. We also set it when loading in arrays from memmap files.

Reply:
Define "annoying" 😀  IMO it's "annoying" to need to reason about mutation and state transitions instead of having things being passed by value and being able to read the code and instantly be able to understand the logical flow 🙃